### PR TITLE
[Tensor] Add `transpose` with out

### PR DESCRIPTION
**Pull request by zhoonit**
- [Tensor] Add `transpose_i` 

```
This patch creates transpose_i, transposes(out) to manipulate given tensor
With exhaustive list of test

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/tensor/tensor.h [hunk 1]
```diff
@@ -504,7 +504,22 @@ class Tensor {
...
+   * @param direction to transpose ex) 0:2:1
+   * @param[out] Tensor to saved to
+   * @retval Tensor& reference to the out
+   */
+  Tensor &transpose(const std::string &direction, Tensor &out) const;
+
+  /**
+   * @brief Transpose a tensor and put it back to @a this
+   *
+   * @param direction to transpose ex) 0:2:1
```
Comment (kparichay) [inline nntrainer/tensor/tensor.h:520]: i highly recommend you to add a note here, that this is NOT an in-place operation. It just returns the result in-place.

File: nntrainer/tensor/tensor.h [hunk 2]
```diff
@@ -504,7 +504,22 @@ class Tensor {
...
+   */
+  void transpose_i(const std::string &direction);
```
Comment (kparichay) [inline nntrainer/tensor/tensor.h:522]: i highly recommend you to add a note here, that this is NOT an in-place operation. It just returns the result in-place.

Comment (kparichay) [reply nntrainer/tensor/tensor.h:522]: infact, unless really needed, lets make this private. This gives a false notion of in-place transpose operation while is it not.

Comment (zhoonit) [reply nntrainer/tensor/tensor.h:522]: Transpose cannot be non-copyable, from behavior point of view, it is what `_i` does. Can you suggest different name?

Comment (kparichay) [reply nntrainer/tensor/tensor.h:522]: How about allow create `Tensor &Tensor::transpose(const std::string &direction, Tensor &out)`, where out can be same as `this`, and this function can call `transpose_i` internally.

**Comment by kparichay** [comment]
Please consider rebasing this PR directly to main, as IMO, it does not depend on other PRs and can be merged faster.

File: nntrainer/tensor/tensor.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.h:522]: Sure, I'll change that way.

Comment (zhoonit) [reply nntrainer/tensor/tensor.h]: Done in a new commit. removed `transpose_i`

**Comment by zhoonit** [comment]
Closing this to try automated merge pr.

**Pull request closed without merging**